NASA-AMMOS/aerie
# Track constraint checking

Closes #997
Closes #514 
Closes #637 

* **Review:** By file
* **Merge strategy:** Merge (no squash)

## Description
This PR adds constraint run tracking. The way it's currently working is each `constraint_run` record is tied to a constraint that was checked. Regardless on if that constraint had violations or not we save a record to the `constraint_run` table that includes basic constraint information and the violations that were encountered. This allows us to reuse those violations in subsequent constraint violation checks.

There is also staleness checking for the runs through a field called `definition_outdated`. This field tracks if the constraint definition has changed since the violations were checked last.

~~There is also staleness checking for the runs, each record gets the status `resolved` when it has been saved to the db. If the user makes a change to a constraint with any `resolved` runs, those affected runs get a status change to `constraint-outdated`. This works the same way for simulations, if a violation check happened and then the user runs a new simulation all constraint runs for that previous dataset are marked as `simulation-outdated`.~~

This is my first big PR for `aerie-merlin` so please hit me with some feedback!

## Verification
Manual tests, I need to work on adding some real tests but I wanted to get the PR up for review first. Expect another commit with tests.

To verify everything is working properly you can do the following:
1. Create a model and a plan.
Add a constraint to the plan (I used `return Real.Resource('/peel').greaterThan(x);` for a lot of my testing).
3. Add an activity to the plan and simulate.
4. Do a constraint violation check, the `constraint_run` table now should be populated with a single entry and it shouldn't be marked as outdated.
5. Go modify your constraint, that entry in the `definition_outdated` field should now be `true`.
6. Do another constraint violation check, the entry should be marked as not outdated again.

~~1. Create a model and a plan.
2. Add a constraint to the plan (I used `return Real.Resource('/peel').greaterThan(x);` for a lot of my testing).
3. Add an activity to the plan and simulate.
4. Do a constraint violation check, the `constraint_run` table now should be populated with a single entry with a status `resolved`.
7. Go modify your constraint, that entry in the db should be now marked as `constraint-outdated`.
8. Do another constraint violation check, a new entry should exist in the table with a status `resolved`.
9. Add another activity and re-simulate.
10. The newest entry in the `constraint_run` table should now be marked `simulation-outdated`.~~

You can also run the constraint violation check in succession to test the caching behavior.

## Documentation
I'll update aerie docs to add information about constraint run tracking after this PR is created. Link will be added here.

## Future work
Going to work on #514 as a separate ticket.
